jmclient/jmclient/wallet_rpc.py

```diff
@@ -1023,3 +1047,111 @@ def getseed(self, request, walletname):
                 raise InvalidRequestFormat()
             seedphrase, _ = self.services["wallet"].get_mnemonic_words()
             return make_jmwalletd_response(request, seedphrase=seedphrase)
+
+        @app.route('/wallet/<string:walletname>/taker/schedule', methods=['POST'])
+        def start_tumbler(self, request, walletname):
+            self.check_cookie(request)
+
+            if not self.services["wallet"]:
+                raise NoWalletFound()
+            if not self.wallet_name == walletname:
+                raise InvalidRequestFormat()
+
+            # -- Options parsing -----------------------------------------------
+
+            (options, args) = get_tumbler_parser().parse_args([])
```

Comment:
Of course, totally understand why we're starting with defaults only. Reviewing the defaults in `jmclient.cli_options.py`, I see only one default that could cause a problem: the default starting mixdepth is 0, but iirc this will raise an error if that mixdepth is empty.

Reply:
I suggest in a future commit/PR: add code to dynamically choose the lowest non-empty mixdepth and set that as `-m`.
However there is a bit of a can of worms there, so as much as that might seem like the *right* thing to do, we might be better off for now instructing the client to only use this algorithm with a non-empty 0 mixdepth, and also instructing them that if they don't have at least three utxos in there, they shouldn't run this.

This might seem a bit weird and janky, and I would not disagree!, but see notes [here](https://github.com/JoinMarket-Org/joinmarket-clientserver/blob/master/docs/SOURCING-COMMITMENTS.md) about the 3 utxos thing. The range-of-mixdepths thing is something I think just needs fixing up.

If users stick to those restrictions, things work with a lot less hassle. If those restrictions are too restrictive (they are!), we just have to improve stuff. People on CLI can kind of fiddle around with the state of the wallet more easily.

Reply:
We are somewhat aware of these kinds of restrictions, which is part of the reason why we default to mixdepth 0 in the Jam UI when depositing funds. Using zero and enabling the scheduler only if 3 UTXOs are present might just work.

I think the way forward is to make it work as best as we can with what we have right now, and improve various things as we go along. The challenge will be to figure out the step-wise increments and in what order to implement them, but I'm weirdly optimistic! 😅

Reply:
> the default starting mixdepth is 0, but iirc this will raise an error if that mixdepth is empty.

In my tests it would still work with a zero mixdepth. For example, with 3 UTXOs on mixdepth 1 and none on mixdepth 0, I got the following schedule:

```
1,0,1,INTERNAL,0.17,16,0
2,0,1,INTERNAL,1.49,16,0
3,0.013100934001158193,1,INTERNAL,1.85,1,0
3,0,1,INTERNAL,0.83,16,0
4,0,1,bcrt1q8urcl2kn925t8zd53gvd2rm5yexeel0t4gedxw,1.84,16,0
```

Note that I used non-default options for faster testing, but `-m` was definitely zero. If I read it correctly, the dynamic selection of the lowest filled mixdepth is [already integrated into the code that generates the schedule](https://github.com/JoinMarket-Org/joinmarket-clientserver/blob/master/jmclient/jmclient/schedule.py#L113-L114). If I'm wrong (which can definitely be the case) I'll add the dynamic mixdepth-selection either to the API function or to the web client. Just let me know.

<details>
<summary>
Click to view the tumbler options used to generate the above schedule.
</summary>
{'mixdepthsrc': 0, 'restart': False, 'schedulefile': 'TUMBLE.schedule', 'addrcount': 1, 'makercountrange': (1, 1), 'minmakercount': 1, 'mixdepthcount': 3, 'txcountparams': (1, 1), 'mintxcount': 1, 'donateamount': 0, 'timelambda': 1, 'stage1_timelambda_increase': 1, 'waittime': 1, 'mincjamount': 100000, 'liquiditywait': 10, 'maxbroadcasts': 4, 'maxcreatetx': 9, 'amtmixdepths': -1, 'rounding_chance': 0.25, 'rounding_sigfig_weights': (55, 15, 25, 65, 40), 'datadir': '', 'recoversync': False, 'wallet_password_stdin': False, 'txfee': -1, 'max_cj_fee_abs': None, 'max_cj_fee_rel': None, 'order_choose_fn': <function fidelity_bond_weighted_order_choose at 0xffff8f7dbc10>
</details>

Reply:
Also good point on the commitment utxos! I'll make sure that this is reflected in the UI flows for the web-client. I feel like this is something we can manage relatively well in the UI (e.g. provide 3 addresses by default in the "Receive" screen when funding the wallet).

Reply:
> In my tests it would still work with a zero mixdepth. For example, with 3 UTXOs on mixdepth 1 and none on mixdepth 0, I got the following schedule:

Yes, you're quite right, I failed to remember it; what you linked is for the initial sweeps, but there is also:

https://github.com/JoinMarket-Org/joinmarket-clientserver/blob/f93ccf0f0c038402ccbcd41f3ddb6eaa21d2b553/jmclient/jmclient/schedule.py#L157

for the non-sweeps, which also respects the lowest initially filled mixdepth.

It would be a really easy and nice PR for someone to add a series of tests of this algo to `jmclient.test.test_schedule.test_tumble_schedule` (which is basically just a stub atm) to ensure it's doing something we all understand :) Your use of 1,1 here obviously fits after thinking about it but it isn't so obvious with that low number :)

Also it's still the case that if someone starts by using `mixdepthcount=10` or something, things go a bit wrong. The reason is that 'we default to 5 mixdepths' has been a bit of a gray area; people have in the past used tumbler through a large number of mixdepths and the code *can* allow this but requires creating a larger wallet. Right now let's please prevent users from trying this unless someone puts in the work to make sure it's all functioning right (btw `amtmixdepths` is directly related, as you could guess).